File: src/test/java/dev/kensa/sentence/scanner/TokenScannerTest.java
package dev.kensa.sentence.scanner;

import dev.kensa.sentence.Token;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

import static dev.kensa.sentence.Token.Type.HighlightedWord;
import static dev.kensa.sentence.Token.Type.Word;
import static java.util.Collections.emptySet;
import static java.util.stream.Collectors.toList;
import static org.assertj.core.api.Assertions.assertThat;

class TokenScannerTest {

    @Test
    void recognisesKeywordsAtStartOfSentenceOnly() {
        var string = "givenSomethingWasGivenWhenThen";

        Indices indices = tokenScannerWith(emptySet()).scan(string);

        indices.stream()
               .filter(index -> index.type() == Token.Type.Keyword)
               .forEach(index -> assertThat(index.start()).isEqualTo(0));
    }

    @Test
    void scansSimpleStringWithNoAcronyms() {
        List<String> expected = List.of("given", "This", "And", "That");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(emptySet()).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void scansStringWithSingleAcronymInMiddle() {
        Set<String> acronyms = Set.of("FTTC", "FTTP", "TT", "BT");

        List<String> expected = List.of("given", "FTTC", "And", "This", "And", "That");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronyms).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    // Kensa#6
    @Test
    void scansStringWithSingleCharacterFollowedByAcronym() {
        Set<String> acronym = Set.of("FTTC");

        List<String> expected = List.of("a", "FTTC", "And", "This", "And", "That");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronym).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void scansStringWithSingleAcronymAtEnd() {
        Set<String> acronym = Set.of("FTTC", "FTTP", "TT", "BT");

        List<String> expected = List.of("given", "And", "This", "And", "That", "FTTC");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronym).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void scansStringWithMultipleAcronyms() {
        Set<String> acronym = Set.of("FTTC", "FTTP", "TT", "BT");

        List<String> expected = List.of("given", "And", "FTTP", "This", "And", "That", "FTTC");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronym).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void choosesLongestMatchingAcronym() {
        Set<String> acronyms = Set.of("FTTC", "FTTP", "FT", "BT");

        List<String> expected = List.of("given", "FT", "And", "FTTP", "This", "And", "That", "FTTC");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronyms).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @ParameterizedTest
    @MethodSource("mixedCaseExamples")
    void scansAcronymsCorrectlyWhenMixedCase(List<String> expected) {
        Set<String> acronyms = Set.of("FTTC");

        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronyms).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    private static Stream<List<String>> mixedCaseExamples() {
        return Stream.of(
                List.of("FTTC"),
                List.of("Fttc"),
                List.of("Ftt", "C"),
                List.of("Ft", "Tc"),
                List.of("Ft", "TC"),
                List.of("F", "Ttc"),
                List.of("Ft", "Tc"),
                List.of("fttc")
        );
    }

    @Test
    void scansStringWithAcronymSpanningCamelWords() {
        Set<String> acronyms = Set.of("ONT");

        List<String> expected = List.of("a", "Notification", "Type", "Of");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronyms).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void scansStringWithAcronymSpanningCamelWords_afterInitialCap() {
        Set<String> acronyms = Set.of("ONT");

        List<String> expected = List.of("a", "Contact");
        var string = String.join("", expected);

        Indices indices = tokenScannerWith(acronyms).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
    }

    @Test
    void scansStringWithHighlightedWords() {
        List<String> expected = List.of("An", "Important", "Priority", "Thing", "Notification");

        var string = String.join("", expected);

        Indices indices = tokenScannerWith(emptySet()).scan(string);

        assertThat(transformed(indices, string)).isEqualTo(expected);
        assertThat(indices.stream()).extracting(Index::type).containsExactly(Word, HighlightedWord, HighlightedWord, Word, Word);
    }

    private List<String> transformed(Indices indices, String string) {
        return indices.stream()
                      .map(index -> string.substring(index.start(), index.end()))
                      .collect(toList());
    }

    private TokenScanner tokenScannerWith(Set<String> acronyms) {
        return new TokenScanner(Set.of("Important", "Priority", "No"), Set.of("Given", "When", "Then"), acronyms);
    }
}

File: src/test/java/dev/kensa/sentence/scanner/TokenScannerTest.kt
package dev.kensa.sentence.scanner

import dev.kensa.sentence.Acronym
import dev.kensa.sentence.Dictionary
import dev.kensa.sentence.HighlightedIdentifier
import dev.kensa.sentence.TokenType.Keyword
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

internal class TokenScannerTest {
    private val acronyms = setOf(
        Acronym.of("FTTC", "Fibre To The Cabinet"),
        Acronym.of("FTTP", "Fibre To The Premises"),
        Acronym.of("TT", "Tourist Trophy"),
        Acronym.of("BT", "British Telecom"),
        Acronym.of("FT", "Financial Times"),
        Acronym.of("ONT", "Optical Network Termination")
    )

    @Test
    internal fun recognisesHighlightedIdentifiers() {
        val identifier = "MyIdentifier"

        val (actual, indices) = tokenScannerWith(emptySet(), setOf(HighlightedIdentifier("MyIdentifier"))).scan(identifier)

        actual shouldBe identifier
        transformed(actual, indices) shouldBe listOf(identifier)
    }

    @Test
    fun recognisesKeywordsAtStartOfSentenceOnly() {
        val string = "givenSomethingWasGivenWhenThen"
        val (_, indices) = tokenScannerWith(emptySet()).scan(string)

        indices.filter { index -> index.type === Keyword }
            .forEach { index -> index.start shouldBe 0 }
    }

    @Test
    internal fun normalisesKotlinWheneverToWhen() {
        val string = "wheneverAThingHappens"
        val expected = listOf("when", "A", "Thing", "Happens")
        val (scanned, indices) = tokenScannerWith(emptySet()).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun splitsInternalNumbersToSeparateTokens() {
        val expected = listOf("last", "24", "Hours")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(emptySet()).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun splitsTrailingNumbersToSeparateTokens() {
        val expected = listOf("last", "24")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(emptySet()).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansSimpleStringWithNoAcronyms() {
        val expected = listOf("given", "This", "And", "That")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(emptySet()).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansStringWithSingleAcronymInMiddle() {
        val expected = listOf("given", "FTTC", "And", "This", "And", "That")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    // Kensa#6
    @Test
    fun scansStringWithSingleCharacterFollowedByAcronym() {
        val expected = listOf("a", "FTTC", "And", "This", "And", "That")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansStringWithSingleAcronymAtEnd() {
        val expected = listOf("given", "And", "This", "And", "That", "FTTC")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansStringWithMultipleAcronyms() {
        val expected = listOf("given", "And", "FTTP", "This", "And", "That", "FTTC")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun choosesLongestMatchingAcronym() {
        val expected = listOf("given", "FT", "And", "FTTP", "This", "And", "That", "FTTC")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @ParameterizedTest
    @MethodSource("mixedCaseExamples")
    fun scansAcronymsCorrectlyWhenMixedCase(expected: List<String>) {
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansStringWithAcronymSpanningCamelWords() {
        val expected = listOf("a", "Notification", "Type", "Of")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    @Test
    fun scansStringWithAcronymSpanningCamelWords_afterInitialCap() {
        val expected = listOf("a", "Contact")
        val string = expected.joinToString("")
        val (scanned, indices) = tokenScannerWith(acronyms).scan(string)

        transformed(scanned, indices) shouldBe expected
    }

    private fun transformed(string: String, indices: Indices): List<String> =
        indices.map { index -> string.substring(index.start, index.end) }.toList()

    private fun tokenScannerWith(acronyms: Set<Acronym>, highlightedIdentifiers: Set<HighlightedIdentifier> = emptySet()): TokenScanner = TokenScanner(
        Dictionary().apply {
            putAcronyms(acronyms)
            putHighlightedIdentifiers(highlightedIdentifiers)
        }
    )

    companion object {
        @JvmStatic
        fun mixedCaseExamples(): Stream<List<String>> {
            return Stream.of(
                listOf("FTTC"),
                listOf("Fttc"),
                listOf("Ftt", "C"),
                listOf("Ft", "Tc"),
                listOf("Ft", "TC"),
                listOf("F", "Ttc"),
                listOf("Ft", "Tc"),
                listOf("fttc")
            )
        }
    }
}